src/adapter/src/coord/sequencer/inner/cluster.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeSet;
+
+use mz_catalog::memory::objects::ClusterVariantManaged;
+use mz_controller::clusters::ReplicaLogging;
+use mz_controller_types::DEFAULT_REPLICA_LOGGING_INTERVAL;
+use mz_ore::instrument;
+use mz_sql::catalog::ObjectType;
+use mz_sql::plan::{AlterClusterPlan, AlterOptionParameter};
+use mz_sql::{plan, session::metadata::SessionMetadata};
+use tracing::Span;
+
+use crate::catalog::ReplicaCreateDropReason;
+use crate::{
+    coord::{AlterCluster, ClusterStage, Coordinator, Message, PlanValidity, StageResult, Staged},
+    session::Session,
+    AdapterError, ExecuteContext, ExecuteResponse,
+};
+
+use super::return_if_err;
+
+impl Staged for ClusterStage {
+    fn validity(&mut self) -> &mut PlanValidity {
+        match self {
+            Self::Alter(stage) => &mut stage.validity,
+        }
+    }
+
+    async fn stage(
+        self,
+        coord: &mut crate::coord::Coordinator,
+        ctx: &mut crate::ExecuteContext,
+    ) -> Result<crate::coord::StageResult<Box<Self>>, crate::AdapterError> {
+        match self {
+            Self::Alter(stage) => {
+                coord
+                    .sequence_alter_cluster(ctx.session(), stage.plan.clone())
+                    .await
+            }
+        }
+    }
+
+    fn message(self, ctx: crate::ExecuteContext, span: tracing::Span) -> crate::coord::Message {
+        Message::ClusterStageReady {
+            ctx,
+            span,
+            stage: self,
+        }
+    }
+
+    fn cancel_enabled(&self) -> bool {
+        // Cluster create and alter are not yet cancelable
+        false
+    }
+}
+
+impl Coordinator {
+    #[instrument]
+    pub(crate) async fn sequence_alter_cluster_staged(
+        &mut self,
+        ctx: ExecuteContext,
+        plan: plan::AlterClusterPlan,
+    ) {
+        let stage = return_if_err!(self.alter_cluster_validate(ctx.session(), plan).await, ctx);
+        self.sequence_staged(ctx, Span::current(), stage).await;
+    }
+
+    #[instrument]
+    async fn alter_cluster_validate(
+        &mut self,
+        session: &Session,
+        plan: plan::AlterClusterPlan,
+    ) -> Result<ClusterStage, AdapterError> {
+        let validity = PlanValidity {
+            transient_revision: self.catalog().transient_revision(),
+            dependency_ids: BTreeSet::new(),
+            cluster_id: None,
```

Comment:
We should add the `cluster_id` from `plan` here. This way if the cluster is concurrently altered in a different way we should bail